src/dimension.c

```diff
@@ -441,3 +519,431 @@ hyperspace_calculate_point(Hyperspace *hs, HeapTuple tuple, TupleDesc tupdesc)
 
 	return p;
 }
+
+static inline int64
+interval_to_usec(Interval *interval)
+{
+	return (interval->month * DAYS_PER_MONTH * USECS_PER_DAY)
+		+ (interval->day * USECS_PER_DAY)
+		+ interval->time;
+}
+
+#define IS_INTEGER_TYPE(type)							\
+	(type == INT2OID || type == INT4OID || type == INT8OID)
+
+#define IS_TIMESTAMP_TYPE(type)									\
+	(type == TIMESTAMPOID || type == TIMESTAMPTZOID || type == DATEOID)
+
+#define IS_VALID_OPEN_DIM_TYPE(type)					\
+	(IS_INTEGER_TYPE(type) || IS_TIMESTAMP_TYPE(type))
+
+#define INT_TYPE_MAX(type)												\
+	(int64)((type == INT2OID) ? INT16_MAX : ((type == INT4OID) ? INT32_MAX : INT64_MAX))
+
+static int64
+get_validated_integer_interval(Oid coltype, int64 value, bool show_warnings)
+{
+	if ((coltype == INT2OID && value > INT16_MAX) ||
+		(coltype == INT4OID && value > INT32_MAX) ||
+		value < 1)
+		ereport(ERROR,
+				(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+				 errmsg("invalid interval: must be between 1 and " INT64_FORMAT,
+						INT_TYPE_MAX(coltype))));
+
+	if (show_warnings && IS_TIMESTAMP_TYPE(coltype) && value < USECS_PER_SEC)
+		ereport(WARNING,
+				(errcode(ERRCODE_AMBIGUOUS_PARAMETER),
+				 errmsg("unexpected interval: smaller than one second"),
+				 errhint("The interval is specified in microseconds")));
+
+	return value;
+}
+
+static int64
+dimension_interval_to_internal(Oid coltype, Oid valuetype, Datum value, bool show_warnings)
```

Comment:
So, there's only one "interval" that this may concern, e.g.:

```
add_dimension('foo', 'time', interval_length => interval '1 day');
```
 Figured it would be clear that it would concern `time` in this case. Space dimensions don't have "intervals". But I can add if you think it makes a big difference.